support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionBuilder.scala

```diff
@@ -4,161 +4,47 @@ import java.util.UUID
 
 import cats.data.EitherT
 import cats.implicits._
-import com.gu.i18n.Country
-import com.gu.monitoring.SafeLogger
 import com.gu.support.catalog.ProductRatePlanId
-import com.gu.support.config.{TouchPointEnvironment, ZuoraDigitalPackConfig}
-import com.gu.support.promotions.{PromoCode, PromoError, PromotionService}
-import com.gu.support.redemption.corporate.CorporateCodeValidator
-import com.gu.support.redemption.gifting.GiftCodeValidator
-import com.gu.support.redemption.gifting.generator.CodeBuilder.GiftCode
-import com.gu.support.redemption.gifting.generator.GiftCodeGeneratorService
-import com.gu.support.redemption.{InvalidCode, InvalidReaderType, ValidCorporateCode}
-import com.gu.support.redemptions.{RedemptionCode, RedemptionData}
-import com.gu.support.workers.ProductTypeRatePlans._
-import com.gu.support.workers.{BillingPeriod, DigitalPack}
-import com.gu.support.zuora.api.ReaderType.{Corporate, Gift}
+import com.gu.support.config.TouchPointEnvironment
+import com.gu.support.promotions.PromoError
+import com.gu.support.redemption.InvalidCode
+import com.gu.support.workers.{DigitalPack, ProductTypeRatePlans}
 import com.gu.support.zuora.api.{ReaderType, SubscriptionData}
-import com.gu.zuora.subscriptionBuilders.ProductSubscriptionBuilders.{applyPromoCodeIfPresent, buildProductSubscription, validateRatePlan}
-import org.joda.time.LocalDate
+import com.gu.zuora.subscriptionBuilders.DigitalSubscriptionBuilder.BuildResult
+import com.gu.zuora.subscriptionBuilders.ProductSubscriptionBuilders.validateRatePlan
 
 import scala.concurrent.{ExecutionContext, Future}
 
-sealed trait SubscriptionPaymentType
+trait SubscriptionPaymentTypeBuilder {
 
-case class SubscriptionPurchase(
-  config: ZuoraDigitalPackConfig,
-  maybePromoCode: Option[PromoCode],
-  billingPeriod: BillingPeriod,
-  country: Country,
-  promotionService: PromotionService
-) extends SubscriptionPaymentType
+  def build(
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    readerType: ReaderType,
+  )(implicit ec: ExecutionContext): BuildResult
 
-case class SubscriptionRedemption(
-  redemptionData: RedemptionData,
-  codeValidator: CorporateCodeValidator
-) extends SubscriptionPaymentType
+}
 
 object DigitalSubscriptionBuilder {
 
   type BuildResult = EitherT[Future, Either[PromoError, InvalidCode], SubscriptionData]
 
   def build(
+    subscriptionPaymentTypeBuilder: SubscriptionPaymentTypeBuilder,
```

Comment:
we lose the pattern match that used to be on line 56-61 and the code is in what used to be the data structure (as it's no longer a data structure)